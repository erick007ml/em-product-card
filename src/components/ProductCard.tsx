import React from 'react'

import { createContext } from 'react'
import { useProduct } from '../hooks/useProduct'
import {
  InitialValues,
  Product,
  ProductCardHandlers,
  ProductContextProps,
} from '../interfaces/productInterfaces'
import styles from '../styles/styles.module.css'

export const ProductContext = createContext({
  product: { id: '', title: '' },
  counter: 0,
  increaseBy: () => {},
} as ProductContextProps)
const { Provider } = ProductContext

export interface OnChangeArgs {
  product: Product
  count: number
}

export interface ProductCardProps {
  product: Product
  // children?: React.ReactElement | React.ReactElement[]
  children: (args: ProductCardHandlers) => JSX.Element
  className?: string
  style?: React.CSSProperties
  onChange?: (args: OnChangeArgs) => void
  value?: number
  initialValues?: InitialValues
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const {
    counter,
    increaseBy,
    maxCount,
    reset,
    isMaxCountReached,
  } = useProduct({
    onChange,
    product,
    value,
    initialValues,
  })
  return (
    <Provider value={{ product, counter, increaseBy, maxCount }}>
      <div style={style} className={`${styles.productCard} ${className}`}>
        {children({
          product,
          count: counter,
          increaseBy,
          maxCount,
          isMaxCountReached,
          reset,
        })}
      </div>
    </Provider>
  )
}

export default ProductCard //ya no es necesario
