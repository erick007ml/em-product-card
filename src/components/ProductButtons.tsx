import React from 'react';

import { useCallback, useContext } from 'react';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface ProductButtonsProps {
  className?: string;
  style?: React.CSSProperties;
}
export const ProductButtons = ({ className, style }: ProductButtonsProps) => {
  const { increaseBy, counter, maxCount } = useContext(ProductContext);

  const isMaxReached = useCallback(() => !!maxCount && counter === maxCount, [
    counter,
    maxCount,
  ]);

  return (
    <div style={style} className={`${styles.buttonsContainer} ${className}`}>
      <button onClick={() => increaseBy(-1)} className={styles.buttonMinus}>
        -
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        onClick={() => increaseBy(+1)}
        className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`}
      >
        +
      </button>
    </div>
  );
};
