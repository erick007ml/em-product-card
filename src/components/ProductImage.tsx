import React from 'react';

import { useContext } from 'react';
import noImage from '../assets/no-image.jpg';
import styles from '../styles/styles.module.css';
import { ProductContext } from './ProductCard';

export interface ProductImageProps {
  img?: string;
  className?: string;
  style?: React.CSSProperties;
}

export const ProductImage = ({
  img = '',
  className,
  style,
}: ProductImageProps) => {
  const { product } = useContext(ProductContext);
  const imgToShow = img || product.img || noImage;

  return (
    <img
      style={style}
      src={imgToShow}
      alt="Product Image"
      className={`${styles.productImg} ${className}`}
    />
  );
};
