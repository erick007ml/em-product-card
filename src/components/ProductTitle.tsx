import React from 'react';

import { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface ProductTitleProps {
  title?: string;
  className?: string;
  style?: React.CSSProperties;
}

export const ProductTitle = ({
  title,
  className,
  style,
}: ProductTitleProps) => {
  const { product } = useContext(ProductContext);
  return (
    <span style={style} className={`${styles.productDescription} ${className}`}>
      {title ? title : product.title}
    </span>
  );
};
