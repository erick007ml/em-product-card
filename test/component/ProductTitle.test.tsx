import React from 'react'
import renderer from 'react-test-renderer'
import { ProductCard, ProductTitle } from '../../src/components'

import { product1, product2 } from '../data/products'

describe('ProductTitle', () => {
  test('debe de mostrar el componente correctamente con el titulo', () => {
    const wrapper = renderer.create(<ProductTitle title={'Product Title'} />)
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
  test('debe de mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>
        {() => (
          <>
            <ProductTitle />
          </>
        )}
      </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
  test('debe de mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product2}>
        {() => (
          <>
            <ProductTitle />
          </>
        )}
      </ProductCard>
    )
    expect(wrapper.toJSON()).toMatchSnapshot()
  })
})
