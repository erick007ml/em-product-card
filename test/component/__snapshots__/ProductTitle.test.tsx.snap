// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Coffee Mug - Card
  </span>
</div>
`;

exports[`ProductTitle debe de mostrar el componente con el nombre del producto 2`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    Pizza - Card
  </span>
</div>
`;

exports[`ProductTitle debe de mostrar el componente correctamente con el titulo 1`] = `
<span
  className="productDescription undefined"
>
  Product Title
</span>
`;
